pkg/url/url.go

```diff
@@ -880,13 +884,30 @@ func Push(client *occlient.Client, kClient *kclient.Client, parameters PushParam
 		urls := parameters.EnvURLS
 		for _, url := range urls {
 			if url.Kind != envinfo.DOCKER {
+				if parameters.EndpointMap == nil {
+					klog.V(4).Infof("No Endpoint entry defined in devfile.")
+					return nil
+				}
+				endpoint, exist := parameters.EndpointMap[int32(url.Port)]
+				if !exist || endpoint.Exposure == "none" || endpoint.Exposure == "internal" {
+					return fmt.Errorf("port %v defined in env.yaml file for URL %v is not exposed in devfile Endpoint entry", url.Port, url.Name)
+				}
+				secure := false
+				if endpoint.Secure || endpoint.Protocol == "https" || endpoint.Protocol == "wss" {
+					secure = true
+				}
+				path := "/"
+				if endpoint.Path != "" {
+					path = endpoint.Path
```

Comment:
This will convert all the URLs on the same port as secure or not secure and based on the one defined in the devfile. It will also use the same path for all the URLs. I think the key to the endpoint map should be the endpoint name.

Reply:
I understand your concern. 
When we only have env.yaml to store URL settings, we expect the devfile not to contain multiple identical ports. It is because the only value we care about in devfile.yaml endpoint entries was `Name` and `TargetPort`. We use these two value to define containers in deployment spec and exposed port in service spec, that is why we expect for a single devfile, there should not exist multiple `Targetport`. 


However, with devfile 2.0, the URL is a little bit tricky now since both env.yaml and endpoints in devfile.yaml contains URL settings and are able to create URL during push. 
If we want to support duplicate port in devfile.yaml in this PR, we encontereed more issues. For example, if multiple endpoints with same `TargetPort` defined in devfile, but with different secure value/path/expose defined, when user call `odo url create` with a url name not defined in devfile, which endpoint entry should it map with?



@girishramnani @kadel @elsony we discussed this issue in last Thursday's meeting, and agreed it cannot be simply done in this PR, and requires more consideration, so we just continue to only allow one single `TargetPort` in this PR. 

We need https://github.com/openshift/odo/issues/3710 and https://github.com/openshift/odo/issues/3718 to address this issue.

Reply:
@mik-dass Do you think that it would make sense to address this in separate PR as part of addressing #3718 ?

Reply:
> For example, if multiple endpoints with same TargetPort defined in devfile, but with different secure value/path/expose defined, when user call odo url create with a url name not defined in devfile, which endpoint entry should it map with?

As you said, we don't support multiple endpoints with the same TargetPort currently in this PR, so it will fail at the time of validation. My concern is why are we deciding the path, secure etc on the basis of the endpoint with the same target port. We currently don't have a flag for path, so it might confuse the user why the URL he/she created is using a different path. I think we should leave the secure, path etc values  for user created URLs to their defaults e.g secure:false, path:/ etc for now.

> @girishramnani @kadel @elsony we discussed this issue in last Thursday's meeting, and agreed it cannot be simply done in this PR, and requires more consideration, so we just continue to only allow one single TargetPort in this PR.

I agree we should do it in a different PR.

Reply:
As I said, I totally understand your concern, how we handle the URL creation is not the best attempt, that is why we need #3718 . And this PR blocks #3718 since it implements the URL creation using devfile.
If we want to implement #3718 as the next PR, does it worth the effort to make this change? As we eventually would have all info saved in devfile.

Reply:
> If we want to implement #3718 as the next PR, does it worth the effort to make this change? As we eventually would have all info saved in devfile.

I know we need this to move forward. But if https://github.com/openshift/odo/issues/3718 doesn't get in time, the migration later might become difficult.

Reply:
Anyways leave it, let's fix them in #3718.

Reply:
> I know we need this to move forward. But if #3718 doesn't get in time, the migration code later to move env URLs to devfile might become tricky.

Agreed. We need to make #3718 a high priority. We should not release 2.0 GA without it